### PR TITLE
YJIT: Stop incrementing chain_depth on defer_compilation

### PR DESCRIPTION
**Pull request by k0kubun**
This PR proposes to isolate a bit for double-defer checks as discussed in https://github.com/ruby/ruby/pull/9426#discussion_r1445028453. 

Currently, `defer_compilation` increments `chain_depth` to check if the block is doubly deferred. It results in confusion like this https://github.com/ruby/ruby/pull/9426#discussion_r1445013315, and unnecessarily decrements the number of chains allowed for the instruction from the max depth constant, which is surprising. As I'm working on megamorphic callsites, this has been too confusing for me.

We use up to 20 chains in a single instruction, so reducing the max chain_depth from 127 to 63 seems fine.

### CODE REVIEW

File: yjit/src/core.rs [hunk 1]
```diff
@@ -1704,30 +1712,39 @@ impl Context {
...
-        self.chain_depth_return_landing &= 0x7f;
+        self.chain_depth_return_landing_defer &= !RETURN_LANDING_BIT;
     }
 
     pub fn is_return_landing(&self) -> bool {
-        self.chain_depth_return_landing & 0x80 > 0
+        self.chain_depth_return_landing_defer & RETURN_LANDING_BIT != 0
+    }
+
+    pub fn set_as_deferred(&mut self) {
```
Comment (maximecb) [inline yjit/src/core.rs:1742]:
```suggestion
    pub fn mark_as_deferred(&mut self) {
```

File: yjit/src/core.rs [hunk 2]
```diff
@@ -456,10 +461,10 @@ pub struct Context {
...
 
     /// Fields packed into u8
-    /// - Lower 7 bits: Depth of this block in the sidechain (eg: inline-cache chain)
-    /// - Top bit: Whether this code is the target of a JIT-to-JIT Ruby return
-    ///   ([Self::is_return_landing])
-    chain_depth_return_landing: u8,
+    /// - 1st bit from the left: Whether this code is the target of a JIT-to-JIT Ruby return ([Self::is_return_landing])
+    /// - 2nd bit from the left: Whether the compilation of this code has been deferred ([Self::is_deferred])
+    /// - Last 6 bits (max: 63): Depth of this block in the sidechain (eg: inline-cache chain)
+    chain_depth_return_landing_defer: u8,
```
Comment (maximecb) [inline yjit/src/core.rs:467]: May want to rename this field to just flags or something like that?

**Review by maximecb — APPROVED** [review]

File: yjit/src/core.rs (hunk 2 above)
Comment (k0kubun) [reply yjit/src/core.rs]: How about `chain_depth_and_flags`? https://github.com/ruby/ruby/pull/9597/commits/f2113c4df37ed69d86a6f72c4f71e3d63a93673e

Comment (maximecb) [reply yjit/src/core.rs]: 👌

**Pull request merged**